File: server/shared/models/workspace.py
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from shared.database import Base
from .user import _nanoid


class Workspace(Base):
    __tablename__ = "workspaces"

    id: Mapped[str] = mapped_column(String(21), primary_key=True, default=_nanoid)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    slug: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(String(500))
    type: Mapped[str] = mapped_column(
        Enum("local", "remote", "cloud", name="workspace_type"), default="local", nullable=False
    )
    path: Mapped[str | None] = mapped_column(String(500))
    owner_id: Mapped[str] = mapped_column(
        String(21), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    settings: Mapped[str] = mapped_column(Text, default="{}", nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )


class WorkspaceMember(Base):
    __tablename__ = "workspace_members"

    workspace_id: Mapped[str] = mapped_column(
        String(21), ForeignKey("workspaces.id", ondelete="CASCADE"), primary_key=True
    )
    user_id: Mapped[str] = mapped_column(
        String(21), ForeignKey("users.id", ondelete="CASCADE"), primary_key=True
    )
    role: Mapped[str] = mapped_column(
        Enum("owner", "admin", "editor", "viewer", name="workspace_member_role"),
        default="viewer",
        nullable=False,
    )
    joined_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: src/services/workspace.ts
import { api } from "./api";

export interface WorkspaceInfo {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  type: "local" | "remote" | "cloud";
  path: string | null;
  owner_id: string;
}

export interface WorkspaceListOut {
  workspaces: WorkspaceInfo[];
  active_workspace_id: string | null;
}

export interface WorkspaceMcpServer {
  id: string;
  name: string;
  url: string | null;
  transport: string;
  workspace_id: string;
}

export interface WorkspaceMcpServerListOut {
  servers: WorkspaceMcpServer[];
}

export const workspaceApi = {
  list() {
    return api.get<WorkspaceListOut>("/workspaces/");
  },

  get(wsId: string) {
    return api.get<WorkspaceInfo>(`/workspaces/${wsId}`);
  },

  listMcpServers(workspaceId: string) {
    return api.get<WorkspaceMcpServerListOut>(`/workspaces/${workspaceId}/mcp-servers`);
  },
};
